// @vkontakte/vkui/dist/components/NativeSelect/NativeSelect.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { getFormFieldModeFromSelectType } from "../../lib/select.js";
import { useIsomorphicLayoutEffect } from "../../lib/useIsomorphicLayoutEffect.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { DropdownIcon } from "../DropdownIcon/DropdownIcon.js";
import { FormField } from "../FormField/FormField.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { SelectTypography } from "../SelectTypography/SelectTypography.js";
const warn = warnOnce('NativeSelect');
const sizeYClassNames = {
    none: "vkuiSelect__sizeYNone",
    compact: "vkuiSelect__sizeYCompact"
};
export const NOT_SELECTED = {
    NATIVE: '__vkui_internal_Select_not_selected__',
    CUSTOM: null
};
/**
 * @visibleName NativeSelect
 */ export const remapFromSelectValueToNativeValue = (value)=>value === NOT_SELECTED.CUSTOM ? NOT_SELECTED.NATIVE : value;
export const remapFromNativeValueToSelectValue = (value)=>value === NOT_SELECTED.NATIVE ? NOT_SELECTED.CUSTOM : value;
/**
 * @see https://vkui.io/components/native-select
 */ export const NativeSelect = (_param)=>{
    var { style: rootStyle, className: rootClassName, getRootRef: rootGetRootRef, align, placeholder, children, getRef, multiline, selectType = 'default', status, icon = /*#__PURE__*/ _jsx(DropdownIcon, {}), before, onChange, value, defaultValue, slotProps } = _param, restProps = _object_without_properties(_param, [
        "style",
        "className",
        "getRootRef",
        "align",
        "placeholder",
        "children",
        "getRef",
        "multiline",
        "selectType",
        "status",
        "icon",
        "before",
        "onChange",
        "value",
        "defaultValue",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ select: { getRootRef: ... } }`');
    }
    const [title, setTitle] = React.useState('');
    const [empty, setEmpty] = React.useState(false);
    const { sizeY = 'none' } = useAdaptivity();
    const _useMergeProps = useMergeProps({
        style: rootStyle,
        className: rootClassName,
        getRootRef: rootGetRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root), { className, style, getRootRef } = _useMergeProps, rootRest = _object_without_properties(_useMergeProps, [
        "className",
        "style",
        "getRootRef"
    ]);
    const _useMergeProps1 = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.select), { getRootRef: getSelectRef } = _useMergeProps1, selectRest = _object_without_properties(_useMergeProps1, [
        "getRootRef"
    ]);
    const selectRef = useExternRef(getSelectRef);
    const checkSelectedOption = ()=>{
        var _selectRef_current;
        const selectedOption = (_selectRef_current = selectRef.current) === null || _selectRef_current === void 0 ? void 0 : _selectRef_current.options[selectRef.current.selectedIndex];
        if (selectedOption) {
            setTitle(selectedOption.text);
            setEmpty(selectedOption.value === NOT_SELECTED.NATIVE && placeholder != null);
        }
    };
    const _onChange = (e)=>{
        const newValue = remapFromNativeValueToSelectValue(e.target.value);
        if (e.target.value === NOT_SELECTED.NATIVE) {
            e.target.value = '';
        }
        if (e.currentTarget.value === NOT_SELECTED.NATIVE) {
            e.currentTarget.value = '';
        }
        onChange === null || onChange === void 0 ? void 0 : onChange(e, newValue);
    };
    useIsomorphicLayoutEffect(checkSelectedOption, [
        children
    ]);
    return /*#__PURE__*/ _jsxs(FormField, _object_spread_props(_object_spread({
        Component: "div",
        className: classNames("vkuiSelect__host", 'vkuiInternalNativeSelect', before && "vkuiSelect__hasBefore", empty && "vkuiSelect__empty", multiline && "vkuiSelect__multiline", align === 'center' && "vkuiSelect__alignCenter", align === 'right' && "vkuiSelect__alignRight", sizeY !== 'regular' && sizeYClassNames[sizeY], className),
        style: style,
        getRootRef: getRootRef,
        disabled: selectRest.disabled,
        before: before,
        after: icon,
        status: status,
        mode: getFormFieldModeFromSelectType(selectType)
    }, rootRest), {
        children: [
            /*#__PURE__*/ _jsxs(RootComponent, _object_spread_props(_object_spread({
                Component: "select",
                baseClassName: "vkuiSelect__el",
                value: value !== undefined ? remapFromSelectValueToNativeValue(value) : value,
                defaultValue: defaultValue !== undefined ? remapFromSelectValueToNativeValue(defaultValue) : defaultValue,
                onChange: callMultiple(_onChange, checkSelectedOption),
                getRootRef: selectRef
            }, selectRest), {
                children: [
                    placeholder && /*#__PURE__*/ _jsx("option", {
                        value: NOT_SELECTED.NATIVE,
                        children: placeholder
                    }),
                    children
                ]
            })),
            /*#__PURE__*/ _jsx("div", {
                className: "vkuiSelect__container",
                "aria-hidden": true,
                children: /*#__PURE__*/ _jsx(SelectTypography, {
                    className: "vkuiSelect__title",
                    selectType: selectType,
                    children: title
                })
            })
        ]
    }));
};
