// @vkontakte/vkui/dist/cssm/components/Input/Input.js
'use client';
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { FormField } from "../FormField/FormField.js";
import { UnstyledTextField } from "../UnstyledTextField/UnstyledTextField.js";
import styles from "./Input.module.css";
const warn = warnOnce('Input');
const sizeYClassNames = {
    none: styles.sizeYNone,
    compact: styles.sizeYCompact
};
/**
 * @see https://vkui.io/components/input
 */ export const Input = ({ type = 'text', align = 'left', getRef, className: rootClassName, getRootRef, style, before, after, status, mode, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const { sizeY = 'none' } = useAdaptivity();
    const { className, ...rootProps } = useMergeProps({
        className: rootClassName,
        getRootRef,
        style
    }, slotProps?.root);
    const inputRest = useMergeProps({
        className: styles.el,
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    return /*#__PURE__*/ _jsx(FormField, {
        className: classNames(styles.host, align === 'right' && styles.alignRight, align === 'center' && styles.alignCenter, sizeY !== 'regular' && sizeYClassNames[sizeY], before && styles.hasBefore, after && styles.hasAfter, className),
        before: before,
        after: after,
        disabled: inputRest.disabled,
        mode: mode,
        status: status,
        ...rootProps,
        children: /*#__PURE__*/ _jsx(UnstyledTextField, {
            as: "input",
            type: type,
            ...inputRest
        })
    });
};
