// @vkontakte/vkui/dist/components/Checkbox/CheckboxSimple/CheckboxSimple.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx } from "react/jsx-runtime";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../../hooks/useAdaptivity.js";
import { useMergeProps } from "../../../hooks/useMergeProps.js";
import { Tappable } from "../../Tappable/Tappable.js";
import { CheckboxInput } from "../CheckboxInput/CheckboxInput.js";
const sizeYClassNames = {
    none: "vkuiCheckboxSimple__sizeYNone",
    compact: "vkuiCheckboxSimple__sizeYCompact"
};
export function CheckboxSimple(_param) {
    var { children, className, style, getRootRef, getRef, description, hoverMode: hoverModeProp, activeMode: activeModeProp, hasHover, hasActive, focusVisibleMode, titleAfter, noPadding, indeterminate, defaultIndeterminate, IconOnCompact, IconOnRegular, IconOffCompact, IconOffRegular, IconIndeterminate, slotProps } = _param, restProps = _object_without_properties(_param, [
        "children",
        "className",
        "style",
        "getRootRef",
        "getRef",
        "description",
        "hoverMode",
        "activeMode",
        "hasHover",
        "hasActive",
        "focusVisibleMode",
        "titleAfter",
        "noPadding",
        "indeterminate",
        "defaultIndeterminate",
        "IconOnCompact",
        "IconOnRegular",
        "IconOffCompact",
        "IconOffRegular",
        "IconIndeterminate",
        "slotProps"
    ]);
    const rootRest = useMergeProps({
        className,
        style,
        getRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const inputRest = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input);
    const { sizeY = 'none' } = useAdaptivity();
    const hoverMode = hoverModeProp || (noPadding ? 'opacity' : 'background');
    const activeMode = activeModeProp || (noPadding ? 'opacity' : 'background');
    return /*#__PURE__*/ _jsx(Tappable, _object_spread_props(_object_spread({
        baseClassName: classNames("vkuiCheckboxSimple__host", !noPadding && "vkuiCheckboxSimple__withPadding", sizeY !== 'regular' && sizeYClassNames[sizeY]),
        disabled: inputRest.disabled,
        hoverMode: hoverMode,
        activeMode: activeMode,
        hasHover: hasHover,
        hasActive: hasActive,
        focusVisibleMode: focusVisibleMode,
        Component: "label"
    }, rootRest), {
        children: /*#__PURE__*/ _jsx(CheckboxInput, {
            indeterminate: indeterminate,
            defaultIndeterminate: defaultIndeterminate,
            IconIndeterminate: IconIndeterminate,
            IconOnCompact: IconOnCompact,
            IconOnRegular: IconOnRegular,
            IconOffCompact: IconOffCompact,
            IconOffRegular: IconOffRegular,
            slotProps: {
                input: inputRest
            }
        })
    }));
}
