// @vkontakte/vkui/dist/cssm/components/Switch/Switch.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useConfigDirection } from "../../hooks/useConfigDirection.js";
import { useFocusVisible } from "../../hooks/useFocusVisible.js";
import { useFocusVisibleClassName } from "../../hooks/useFocusVisibleClassName.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { VisuallyHidden } from "../VisuallyHidden/VisuallyHidden.js";
import styles from "./Switch.module.css";
const warn = warnOnce('Switch');
const sizeYClassNames = {
    none: styles.sizeYNone,
    compact: styles.sizeYCompact
};
/**
 * @see https://vkui.io/components/switch
 */ export const Switch = ({ style: rootStyle, className: rootClassName, getRootRef: rootGetRootRef, getRef, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const rootRest = useMergeProps({
        style: rootStyle,
        className: rootClassName,
        getRootRef: rootGetRootRef
    }, slotProps?.root);
    const { checked: checkedProp, onBlur: onBlurProp, onFocus: onFocusProp, onClick, ...inputRest } = useMergeProps({
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    const direction = useConfigDirection();
    const isRtl = direction === 'rtl';
    const platform = usePlatform();
    const { sizeY = 'none' } = useAdaptivity();
    const { focusVisible, onBlur, onFocus } = useFocusVisible();
    const focusVisibleClassNames = useFocusVisibleClassName({
        focusVisible,
        mode: 'outside'
    });
    const handleBlur = callMultiple(onBlur, onBlurProp);
    const handleFocus = callMultiple(onFocus, onFocusProp);
    const [localUncontrolledChecked, setLocalUncontrolledChecked] = React.useState(Boolean(inputRest.defaultChecked));
    const isControlled = checkedProp !== undefined;
    const syncUncontrolledCheckedStateOnClick = React.useCallback((e)=>{
        if (isControlled) {
            return;
        }
        const switchTarget = e.target;
        setLocalUncontrolledChecked(switchTarget.checked);
    }, [
        isControlled
    ]);
    const inputProps = {
        Component: 'input',
        type: 'checkbox',
        role: 'switch',
        onBlur: handleBlur,
        onFocus: handleFocus,
        onClick: callMultiple(syncUncontrolledCheckedStateOnClick, onClick),
        ...inputRest
    };
    if (isControlled) {
        inputProps.checked = checkedProp;
        inputProps['aria-checked'] = checkedProp ? 'true' : 'false';
    } else {
        inputProps['aria-checked'] = localUncontrolledChecked ? 'true' : 'false';
    }
    return /*#__PURE__*/ _jsxs(RootComponent, {
        Component: "label",
        baseClassName: classNames(styles.host, sizeY !== 'regular' && sizeYClassNames[sizeY], platform === 'ios' ? styles.ios : styles.default, inputRest.disabled && styles.disabled, isRtl && styles.rtl, focusVisibleClassNames),
        ...rootRest,
        children: [
            /*#__PURE__*/ _jsx(VisuallyHidden, {
                baseClassName: styles.inputNative,
                ...inputProps
            }),
            /*#__PURE__*/ _jsxs("span", {
                "aria-hidden": true,
                className: styles.inputFake,
                children: [
                    /*#__PURE__*/ _jsx("span", {
                        className: styles.track
                    }),
                    /*#__PURE__*/ _jsx("span", {
                        "aria-hidden": true,
                        className: classNames(styles.handle, platform !== 'ios' && !inputRest.disabled && styles.handleWithRipple)
                    })
                ]
            })
        ]
    });
};
