// @vkontakte/vkui/dist/components/Textarea/Textarea.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { useResizeObserver } from "../../hooks/useResizeObserver.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { useDOM } from "../../lib/dom.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { FormField } from "../FormField/FormField.js";
import { UnstyledTextField } from "../UnstyledTextField/UnstyledTextField.js";
import { useResizeTextarea } from "./useResizeTextarea.js";
const warn = warnOnce('Textarea');
const sizeYClassNames = {
    none: "vkuiTextarea__sizeYNone",
    compact: "vkuiTextarea__sizeYCompact"
};
/**
 * @see https://vkui.io/components/textarea
 */ export const Textarea = (_param)=>{
    var { className: rootClassName, getRootRef, style, grow = true, onResize, getRef, rows = 2, maxHeight, status, align, mode, after, before, afterAlign, beforeAlign, slotProps } = _param, restProps = _object_without_properties(_param, [
        "className",
        "getRootRef",
        "style",
        "grow",
        "onResize",
        "getRef",
        "rows",
        "maxHeight",
        "status",
        "align",
        "mode",
        "after",
        "before",
        "afterAlign",
        "beforeAlign",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ textArea: { getRootRef: ... } }`');
    }
    const { sizeY = 'none' } = useAdaptivity();
    const platform = usePlatform();
    const { window } = useDOM();
    const _useMergeProps = useMergeProps({
        className: rootClassName,
        getRootRef,
        style
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root), { className } = _useMergeProps, rootProps = _object_without_properties(_useMergeProps, [
        "className"
    ]);
    const _useMergeProps1 = useMergeProps(_object_spread({
        className: "vkuiTextarea__el",
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.textArea), { onChange, getRootRef: getTextAreaRef, value } = _useMergeProps1, textAreaRest = _object_without_properties(_useMergeProps1, [
        "onChange",
        "getRootRef",
        "value"
    ]);
    const [refResizeTextarea, resize] = useResizeTextarea(onResize, grow);
    const elementRef = useExternRef(getTextAreaRef, refResizeTextarea);
    React.useEffect(resize, [
        resize,
        sizeY,
        platform,
        value
    ]);
    useResizeObserver(window, resize);
    return /*#__PURE__*/ _jsx(FormField, _object_spread_props(_object_spread({
        className: classNames("vkuiTextarea__host", sizeY !== 'regular' && sizeYClassNames[sizeY], align === 'right' && "vkuiTextarea__alignRight", align === 'center' && "vkuiTextarea__alignCenter", className),
        disabled: textAreaRest.disabled,
        status: status,
        mode: mode,
        after: after,
        before: before,
        afterAlign: afterAlign,
        beforeAlign: beforeAlign,
        maxHeight: maxHeight
    }, rootProps), {
        children: /*#__PURE__*/ _jsx(UnstyledTextField, _object_spread({
            value: value,
            as: "textarea",
            rows: rows,
            getRootRef: elementRef,
            onChange: callMultiple(onChange, resize)
        }, textAreaRest))
    }));
};
