// @vkontakte/vkui/dist/cssm/components/Flex/Flex.js
import { jsx as _jsx } from "react/jsx-runtime";
import { classNames } from "@vkontakte/vkjs";
import { calculateGap, columnGapClassNames, resolveLayoutProps, rowGapClassNames } from "../../lib/layouts/index.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { FlexItem } from "./FlexItem/FlexItem.js";
import styles from "./Flex.module.css";
import flexItemStyles from "./FlexItem/FlexItem.module.css";
const justifyClassNames = {
    'start': styles.justifyStart,
    'end': styles.justifyEnd,
    'center': styles.justifyCenter,
    'space-around': styles.justifySpaceAround,
    'space-between': styles.justifySpaceBetween,
    'space-evenly': styles.justifySpaceEvenly
};
const alignClassNames = {
    start: styles.alignStart,
    end: styles.alignEnd,
    center: styles.alignCenter,
    stretch: styles.alignStretch,
    baseline: styles.alignBaseline
};
const alignSelfClassNames = {
    start: flexItemStyles.alignSelfStart,
    end: flexItemStyles.alignSelfEnd,
    center: flexItemStyles.alignSelfCenter,
    baseline: flexItemStyles.alignSelfBaseline,
    stretch: flexItemStyles.alignSelfStretch
};
const displayClassNames = {
    'none': styles.displayNone,
    'inline-flex': styles.displayInlineFlex
};
/**
 * @see https://vkui.io/components/flex
 */ export const Flex = ({ gap = 0, align, justify, margin = 'none', noWrap = false, direction = 'row', reverse = false, children, alignSelf, display = 'flex', ...restProps })=>{
    const [rowGap, columnGap] = calculateGap(gap);
    const resolvedProps = resolveLayoutProps(restProps);
    return /*#__PURE__*/ _jsx(RootComponent, {
        ...resolvedProps,
        baseClassName: classNames(styles.host, !noWrap && styles.wrap, reverse && styles.reverse, direction !== 'row' && styles.directionColumn, margin !== 'none' && styles.marginAuto, align && alignClassNames[align], alignSelf && alignSelfClassNames[alignSelf], justify && justifyClassNames[justify], getGapsPresets(rowGap, columnGap), display !== 'flex' && displayClassNames[display]),
        baseStyle: getGapsByUser(rowGap, columnGap),
        children: children
    });
};
function getGapsPresets(rowGap, columnGap) {
    return classNames(typeof rowGap === 'string' && rowGapClassNames[rowGap], typeof columnGap === 'string' && columnGapClassNames[columnGap]);
}
function getGapsByUser(rowGap, columnGap) {
    const style = {};
    if (typeof rowGap === 'number') {
        style['--vkui_internal--row_gap'] = `${rowGap}px`;
    }
    if (typeof columnGap === 'number') {
        style['--vkui_internal--column_gap'] = `${columnGap}px`;
    }
    return style;
}
Flex.Item = FlexItem;
