// @vkontakte/vkui/dist/cssm/components/File/File.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { Button } from "../Button/Button.js";
import { VisuallyHidden } from "../VisuallyHidden/VisuallyHidden.js";
const warn = warnOnce('File');
/**
 * @see https://vkui.io/components/file
 */ export const File = ({ getRootRef, className, style, children = 'Выберите файл', align = 'left', size, mode, stretched, before, after, loading, getRef, appearance, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const rootProps = useMergeProps({
        className,
        style,
        getRootRef: getRootRef
    }, slotProps?.root);
    const inputRest = useMergeProps({
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    return /*#__PURE__*/ _jsxs(Button, {
        Component: "label",
        align: align,
        stretched: stretched,
        mode: mode,
        appearance: appearance,
        size: size,
        before: before,
        after: after,
        loading: loading,
        disabled: inputRest.disabled,
        ...rootProps,
        children: [
            /*#__PURE__*/ _jsx(VisuallyHidden, {
                title: "",
                Component: "input",
                type: "file",
                ...inputRest
            }),
            children
        ]
    });
};
