// @vkontakte/vkui/dist/components/CarouselBase/hooks.js
import * as React from "react";
import { animate } from "../../lib/animate.js";
import { cubicBezier } from "../../lib/fx.js";
const PRESET_ANIMATION_EASING = {
    'linear': [
        0,
        0,
        1,
        1
    ],
    'ease': [
        0.25,
        0.1,
        0.25,
        1
    ],
    'ease-in': [
        0.42,
        0,
        1,
        1
    ],
    'ease-out': [
        0,
        0,
        0.58,
        1
    ],
    'ease-in-out': [
        0.42,
        0,
        0.58,
        1
    ]
};
export function useSlideAnimation(animationDuration, animationEasing) {
    const animationQueue = React.useRef([]);
    function getAnimateFunction(drawFunction) {
        return ()=>{
            animate({
                duration: animationDuration,
                timing: getAnimationTiming(animationEasing),
                animationQueue: animationQueue.current,
                draw: drawFunction
            });
        };
    }
    function addToAnimationQueue(func) {
        animationQueue.current.push(func);
    }
    function startAnimation() {
        if (animationQueue.current.length === 1) {
            animationQueue.current[0]();
        }
    }
    function animationInQueue() {
        return !!animationQueue.current.length;
    }
    function getAnimationEasing() {
        return `cubic-bezier(${(typeof animationEasing === 'string' ? PRESET_ANIMATION_EASING[animationEasing] : animationEasing).join(', ')})`;
    }
    return {
        animationInQueue,
        getAnimateFunction,
        addToAnimationQueue,
        startAnimation,
        getAnimationEasing
    };
}
function getAnimationTiming(animationEasing) {
    if (typeof animationEasing === 'string') {
        return cubicBezier(...PRESET_ANIMATION_EASING[animationEasing]);
    }
    return cubicBezier(...animationEasing);
}
