// @vkontakte/vkui/dist/cssm/components/WriteBar/WriteBar.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames, hasReactNode } from "@vkontakte/vkjs";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { useResizeTextarea } from "../Textarea/useResizeTextarea.js";
import { Headline } from "../Typography/Headline/Headline.js";
import { Title } from "../Typography/Title/Title.js";
import styles from "./WriteBar.module.css";
const warn = warnOnce('WriteBar');
const WriteBarTypography = (props)=>{
    const platform = usePlatform();
    if (platform === 'ios') {
        return /*#__PURE__*/ _jsx(Title, {
            ...props,
            level: "3",
            weight: "3"
        });
    }
    return /*#__PURE__*/ _jsx(Headline, {
        weight: "3",
        ...props
    });
};
/**
 * @see https://vkui.io/components/write-bar
 */ export const WriteBar = ({ className, style, before, inlineAfter, after, getRootRef, getRef, onHeightChange, shadow = false, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ textArea: { getRootRef: ... } }`');
    }
    const platform = usePlatform();
    const rootProps = useMergeProps({
        className,
        getRootRef,
        style
    }, slotProps?.root);
    const { onChange, getRootRef: getTextAreaRef, ...textAreaRest } = useMergeProps({
        className: styles.textarea,
        getRootRef: getRef,
        ...restProps
    }, slotProps?.textArea);
    const [refResizeTextarea, resize] = useResizeTextarea(onHeightChange, true);
    const textareaRef = useExternRef(getTextAreaRef, refResizeTextarea);
    React.useEffect(resize, [
        resize,
        platform
    ]);
    return /*#__PURE__*/ _jsx(RootComponent, {
        baseClassName: classNames(styles.host, platform === 'ios' && styles.ios, shadow && styles.shadow),
        ...rootProps,
        children: /*#__PURE__*/ _jsxs("div", {
            className: styles.form,
            children: [
                hasReactNode(before) && /*#__PURE__*/ _jsx("div", {
                    className: styles.before,
                    children: before
                }),
                /*#__PURE__*/ _jsxs("div", {
                    className: styles.formIn,
                    children: [
                        /*#__PURE__*/ _jsx(WriteBarTypography, {
                            Component: "textarea",
                            onChange: callMultiple(onChange, resize),
                            getRootRef: textareaRef,
                            ...textAreaRest
                        }),
                        hasReactNode(inlineAfter) && /*#__PURE__*/ _jsx("div", {
                            className: styles.inlineAfter,
                            children: inlineAfter
                        })
                    ]
                }),
                hasReactNode(after) && /*#__PURE__*/ _jsx("div", {
                    className: styles.after,
                    children: after
                })
            ]
        })
    });
};
