// @vkontakte/vkjs/lib/other/index.js
/**
 * classNames
 */ export { clsx as classNames } from "clsx";
/**
 * clipboard
 */ export { copyTextToClipboard } from "./clipboard.js";
/**
 * common
 */ export { isNumeric } from "./common.js";
/**
 * detections
 */ export { isPassiveEventsSupported } from "./detections.js";
export { isSmoothScrollSupported } from "./detections.js";
export { isEqual } from "./equal.js";
export { noop } from "./functions.js";
export { once } from "./functions.js";
export { getCookie } from "./cookie.js";
export { isCookieEnabled } from "./cookie.js";
/**
 * OffsetRect
 */ export { getOffsetRect } from "./getOffsetRect.js";
export { getPhotoSize } from "./getPhotoSize.js";
/**
 * objects
 */ export { deleteObjectKeys } from "./objects.js";
/**
 * querystring
 */ export { querystring } from "./querystring.js";
/**
 * random
 */ export { getRandomInt } from "./random.js";
export { getRandomString } from "./random.js";
export { hasReactNode } from "./react_utils.js";
export { isPrimitiveReactNode } from "./react_utils.js";
export { escapeRegExp } from "./regexp.js";
export { localStorage } from "./storage.js";
export { sessionStorage } from "./storage.js";
export { canUseDOM } from "./dom.js";
export { canUseEventListeners } from "./dom.js";
export { onDOMLoaded } from "./dom.js";
