// @vkontakte/vkui/dist/cssm/components/Textarea/Textarea.js
'use client';
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { useResizeObserver } from "../../hooks/useResizeObserver.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { useDOM } from "../../lib/dom.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { FormField } from "../FormField/FormField.js";
import { UnstyledTextField } from "../UnstyledTextField/UnstyledTextField.js";
import { useResizeTextarea } from "./useResizeTextarea.js";
import styles from "./Textarea.module.css";
const warn = warnOnce('Textarea');
const sizeYClassNames = {
    none: styles.sizeYNone,
    compact: styles.sizeYCompact
};
/**
 * @see https://vkui.io/components/textarea
 */ export const Textarea = ({ className: rootClassName, getRootRef, style, grow = true, onResize, getRef, rows = 2, maxHeight, status, align, mode, after, before, afterAlign, beforeAlign, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ textArea: { getRootRef: ... } }`');
    }
    const { sizeY = 'none' } = useAdaptivity();
    const platform = usePlatform();
    const { window } = useDOM();
    const { className, ...rootProps } = useMergeProps({
        className: rootClassName,
        getRootRef,
        style
    }, slotProps?.root);
    const { onChange, getRootRef: getTextAreaRef, value, ...textAreaRest } = useMergeProps({
        className: styles.el,
        getRootRef: getRef,
        ...restProps
    }, slotProps?.textArea);
    const [refResizeTextarea, resize] = useResizeTextarea(onResize, grow);
    const elementRef = useExternRef(getTextAreaRef, refResizeTextarea);
    React.useEffect(resize, [
        resize,
        sizeY,
        platform,
        value
    ]);
    useResizeObserver(window, resize);
    return /*#__PURE__*/ _jsx(FormField, {
        className: classNames(styles.host, sizeY !== 'regular' && sizeYClassNames[sizeY], align === 'right' && styles.alignRight, align === 'center' && styles.alignCenter, className),
        disabled: textAreaRest.disabled,
        status: status,
        mode: mode,
        after: after,
        before: before,
        afterAlign: afterAlign,
        beforeAlign: beforeAlign,
        maxHeight: maxHeight,
        ...rootProps,
        children: /*#__PURE__*/ _jsx(UnstyledTextField, {
            value: value,
            as: "textarea",
            rows: rows,
            getRootRef: elementRef,
            onChange: callMultiple(onChange, resize),
            ...textAreaRest
        })
    });
};
