// @vkontakte/vkui/dist/components/Checkbox/Checkbox.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { hasReactNode } from "@vkontakte/vkjs";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { SelectionControl } from "../SelectionControl/SelectionControl.js";
import { SelectionControlLabel } from "../SelectionControl/SelectionControlLabel/SelectionControlLabel.js";
import { CheckboxInput } from "./CheckboxInput/CheckboxInput.js";
import { CheckboxSimple } from "./CheckboxSimple/CheckboxSimple.js";
const warn = warnOnce('Checkbox');
const CheckboxComponent = (_param)=>{
    var { children, className, style, getRootRef, getRef, description, hoverMode, activeMode, hasHover, hasActive, focusVisibleMode, titleAfter, noPadding, indeterminate, defaultIndeterminate, IconOnCompact, IconOnRegular, IconOffCompact, IconOffRegular, IconIndeterminate, slotProps } = _param, restProps = _object_without_properties(_param, [
        "children",
        "className",
        "style",
        "getRootRef",
        "getRef",
        "description",
        "hoverMode",
        "activeMode",
        "hasHover",
        "hasActive",
        "focusVisibleMode",
        "titleAfter",
        "noPadding",
        "indeterminate",
        "defaultIndeterminate",
        "IconOnCompact",
        "IconOnRegular",
        "IconOffCompact",
        "IconOffRegular",
        "IconIndeterminate",
        "slotProps"
    ]);
    const rootRest = useMergeProps({
        className,
        style,
        getRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const inputRest = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input);
    return /*#__PURE__*/ _jsxs(SelectionControl, _object_spread_props(_object_spread({
        disabled: inputRest.disabled,
        hoverMode: hoverMode,
        activeMode: activeMode,
        hasHover: hasHover,
        hasActive: hasActive,
        focusVisibleMode: focusVisibleMode,
        noPadding: noPadding
    }, rootRest), {
        children: [
            /*#__PURE__*/ _jsx(CheckboxInput, {
                indeterminate: indeterminate,
                defaultIndeterminate: defaultIndeterminate,
                IconIndeterminate: IconIndeterminate,
                IconOnCompact: IconOnCompact,
                IconOnRegular: IconOnRegular,
                IconOffCompact: IconOffCompact,
                IconOffRegular: IconOffRegular,
                slotProps: {
                    input: inputRest
                }
            }),
            /*#__PURE__*/ _jsx(SelectionControlLabel, {
                titleAfter: titleAfter,
                description: description,
                children: children
            })
        ]
    }));
};
/**
 * @see https://vkui.io/components/checkbox
 */ export const Checkbox = (props)=>{
    if (process.env.NODE_ENV === 'development' && props.getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const simple = !(hasReactNode(props.children) || hasReactNode(props.description));
    if (simple) {
        return /*#__PURE__*/ _jsx(CheckboxSimple, _object_spread({}, props));
    }
    return /*#__PURE__*/ _jsx(CheckboxComponent, _object_spread({}, props));
};
Checkbox.Input = CheckboxInput;
