// @vkontakte/vkui/dist/components/CustomSelect/CustomSelectInput/CustomSelectInput.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable jsdoc/require-jsdoc */ import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../../hooks/useAdaptivity.js";
import { useExternRef } from "../../../hooks/useExternRef.js";
import { useFocusWithin } from "../../../hooks/useFocusWithin.js";
import { useMergeProps } from "../../../hooks/useMergeProps.js";
import { usePlatform } from "../../../hooks/usePlatform.js";
import { getFormFieldModeFromSelectType } from "../../../lib/select.js";
import { FormField } from "../../FormField/FormField.js";
import { SelectTypography } from "../../SelectTypography/SelectTypography.js";
import { Text } from "../../Typography/Text/Text.js";
import { VisuallyHidden } from "../../VisuallyHidden/VisuallyHidden.js";
const sizeYClassNames = {
    none: "vkuiCustomSelectInput__sizeYNone",
    compact: "vkuiCustomSelectInput__sizeYCompact"
};
/**
 * @since 5.10.0
 * @private
 */ export const CustomSelectInput = (_param)=>{
    var { style: rootStyle, className: rootClassName, getRootRef: rootGetRootRef, align = 'left', getRef, before, after, status, children, placeholder, selectType = 'default', multiline, fetching, labelTextTestId, searchable, accessible, slotProps } = _param, restProps = _object_without_properties(_param, [
        "style",
        "className",
        "getRootRef",
        "align",
        "getRef",
        "before",
        "after",
        "status",
        "children",
        "placeholder",
        "selectType",
        "multiline",
        "fetching",
        "labelTextTestId",
        "searchable",
        "accessible",
        "slotProps"
    ]);
    const { sizeY = 'none' } = useAdaptivity();
    const _useMergeProps = useMergeProps({
        style: rootStyle,
        className: rootClassName,
        getRootRef: rootGetRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root), { style, className, getRootRef } = _useMergeProps, rootRest = _object_without_properties(_useMergeProps, [
        "style",
        "className",
        "getRootRef"
    ]);
    const _useMergeProps1 = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input), { className: inputClassName, value, readOnly, disabled } = _useMergeProps1, inputProps = _object_without_properties(_useMergeProps1, [
        "className",
        "value",
        "readOnly",
        "disabled"
    ]);
    const title = children || placeholder;
    const showLabelOrPlaceholder = !Boolean(value);
    const handleRootRef = useExternRef(getRootRef);
    const focusWithin = useFocusWithin(handleRootRef);
    const inputReadonly = readOnly || disabled && fetching;
    const input = /*#__PURE__*/ _jsx(Text, _object_spread({
        type: "text",
        disabled: disabled && !fetching,
        readOnly: inputReadonly,
        Component: "input",
        normalize: false,
        placeholder: children ? '' : placeholder,
        className: classNames("vkuiCustomSelectInput__el", (readOnly || showLabelOrPlaceholder && !focusWithin) && "vkuiCustomSelectInput__elCursorPointer", inputClassName),
        value: value
    }, inputProps));
    const inputHidden = React.useMemo(()=>{
        if (accessible) {
            if (!searchable) {
                return true;
            }
            return !focusWithin || inputReadonly && !fetching;
        } else {
            return false;
        }
    }, [
        accessible,
        fetching,
        focusWithin,
        inputReadonly,
        searchable
    ]);
    const labelHidden = showLabelOrPlaceholder ? false : !inputHidden;
    const platform = usePlatform();
    return /*#__PURE__*/ _jsx(FormField, _object_spread_props(_object_spread({
        Component: "div",
        style: style,
        className: classNames("vkuiCustomSelectInput__host", align === 'right' && "vkuiCustomSelectInput__alignRight", align === 'center' && "vkuiCustomSelectInput__alignCenter", !children && "vkuiCustomSelectInput__empty", multiline && "vkuiCustomSelectInput__multiline", sizeY !== 'regular' && sizeYClassNames[sizeY], before && "vkuiCustomSelectInput__hasBefore", after && "vkuiCustomSelectInput__hasAfter", inputHidden && "vkuiCustomSelectInput__inputHidden", labelHidden && "vkuiCustomSelectInput__labelHidden", accessible && "vkuiCustomSelectInput__accessible", value && "vkuiCustomSelectInput__hasInputValue", className),
        getRootRef: handleRootRef,
        before: before,
        after: after,
        disabled: disabled,
        mode: getFormFieldModeFromSelectType(selectType),
        status: status
    }, rootRest), {
        children: /*#__PURE__*/ _jsxs("div", {
            className: "vkuiCustomSelectInput__inputGroup",
            children: [
                /*#__PURE__*/ _jsx("div", {
                    className: classNames("vkuiCustomSelectInput__container", className),
                    tabIndex: -1,
                    "aria-hidden": true,
                    "data-testid": labelTextTestId,
                    children: /*#__PURE__*/ _jsx(SelectTypography, {
                        selectType: selectType,
                        className: "vkuiCustomSelectInput__title",
                        children: title
                    })
                }),
                readOnly && platform === 'ios' ? /*#__PURE__*/ _jsx(VisuallyHidden, {
                    children: input
                }) : input
            ]
        })
    }));
};
