// @vkontakte/vkui/dist/cssm/components/Checkbox/Checkbox.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { hasReactNode } from "@vkontakte/vkjs";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { SelectionControl } from "../SelectionControl/SelectionControl.js";
import { SelectionControlLabel } from "../SelectionControl/SelectionControlLabel/SelectionControlLabel.js";
import { CheckboxInput } from "./CheckboxInput/CheckboxInput.js";
import { CheckboxSimple } from "./CheckboxSimple/CheckboxSimple.js";
const warn = warnOnce('Checkbox');
const CheckboxComponent = ({ children, className, style, getRootRef, getRef, description, hoverMode, activeMode, hasHover, hasActive, focusVisibleMode, titleAfter, noPadding, indeterminate, defaultIndeterminate, IconOnCompact, IconOnRegular, IconOffCompact, IconOffRegular, IconIndeterminate, slotProps, ...restProps })=>{
    const rootRest = useMergeProps({
        className,
        style,
        getRootRef
    }, slotProps?.root);
    const inputRest = useMergeProps({
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    return /*#__PURE__*/ _jsxs(SelectionControl, {
        disabled: inputRest.disabled,
        hoverMode: hoverMode,
        activeMode: activeMode,
        hasHover: hasHover,
        hasActive: hasActive,
        focusVisibleMode: focusVisibleMode,
        noPadding: noPadding,
        ...rootRest,
        children: [
            /*#__PURE__*/ _jsx(CheckboxInput, {
                indeterminate: indeterminate,
                defaultIndeterminate: defaultIndeterminate,
                IconIndeterminate: IconIndeterminate,
                IconOnCompact: IconOnCompact,
                IconOnRegular: IconOnRegular,
                IconOffCompact: IconOffCompact,
                IconOffRegular: IconOffRegular,
                slotProps: {
                    input: inputRest
                }
            }),
            /*#__PURE__*/ _jsx(SelectionControlLabel, {
                titleAfter: titleAfter,
                description: description,
                children: children
            })
        ]
    });
};
/**
 * @see https://vkui.io/components/checkbox
 */ export const Checkbox = (props)=>{
    if (process.env.NODE_ENV === 'development' && props.getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const simple = !(hasReactNode(props.children) || hasReactNode(props.description));
    if (simple) {
        return /*#__PURE__*/ _jsx(CheckboxSimple, {
            ...props
        });
    }
    return /*#__PURE__*/ _jsx(CheckboxComponent, {
        ...props
    });
};
Checkbox.Input = CheckboxInput;
