// @vkontakte/vkjs/lib/typecheck/type_checkers.js
import { _ as _type_of } from "@swc/helpers/_/_type_of";
export function isObjectLike(object) {
    return (typeof object === "undefined" ? "undefined" : _type_of(object)) === 'object' && object !== null;
}
export function isArray(object) {
    return Array.isArray(object);
}
export function isObject(object) {
    return Object.prototype.toString.call(object) === '[object Object]';
}
export function isUndefined(object) {
    return typeof object === 'undefined';
}
export function isFunction(object) {
    return typeof object === 'function';
}
export function isFormData(object) {
    return object && Object.prototype.toString.call(object) === '[object FormData]';
}
export function isString(object) {
    return typeof object === 'string';
}
/**
 * Проверяет что переданное значение является `Number`
 *
 * ## Пример
 *
 * ```ts
 * import assert from 'node:assert';
 * import { isNumber } from '@vkontakte/vkjs';
 *
 * assert.strictEqual(isNumber(3), true);
 * assert.strictEqual(isNumber(Infinity), true);
 * assert.strictEqual(isNumber('3'), false);
 * ```
 *
 * Для исключения `Infinity` `-Infinity` и `NaN` используйте `Number.isFinite`
 *
 * ```ts
 * import assert from 'node:assert';
 *
 * assert.strictEqual(Number.isFinite(3), true);
 * assert.strictEqual(Number.isFinite(Infinity), false);
 * assert.strictEqual(Number.isFinite('3'), false);
 * ```
 */ export function isNumber(object) {
    return typeof object === 'number';
}
export function isPromiseLike(object) {
    return (isObject(object) || isFunction(object)) && isFunction(object.then);
}
