// @vkontakte/vkui/dist/components/File/File.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { Button } from "../Button/Button.js";
import { VisuallyHidden } from "../VisuallyHidden/VisuallyHidden.js";
const warn = warnOnce('File');
/**
 * @see https://vkui.io/components/file
 */ export const File = (_param)=>{
    var { getRootRef, className, style, children = 'Выберите файл', align = 'left', size, mode, stretched, before, after, loading, getRef, appearance, slotProps } = _param, restProps = _object_without_properties(_param, [
        "getRootRef",
        "className",
        "style",
        "children",
        "align",
        "size",
        "mode",
        "stretched",
        "before",
        "after",
        "loading",
        "getRef",
        "appearance",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const rootProps = useMergeProps({
        className,
        style,
        getRootRef: getRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const inputRest = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input);
    return /*#__PURE__*/ _jsxs(Button, _object_spread_props(_object_spread({
        Component: "label",
        align: align,
        stretched: stretched,
        mode: mode,
        appearance: appearance,
        size: size,
        before: before,
        after: after,
        loading: loading,
        disabled: inputRest.disabled
    }, rootProps), {
        children: [
            /*#__PURE__*/ _jsx(VisuallyHidden, _object_spread({
                title: "",
                Component: "input",
                type: "file"
            }, inputRest)),
            children
        ]
    }));
};
