// @vkontakte/vkui/dist/cssm/components/Flex/FlexItem/FlexItem.js
import { jsx as _jsx } from "react/jsx-runtime";
import { classNames } from "@vkontakte/vkjs";
import { warnOnce } from "../../../lib/warnOnce.js";
import { RootComponent } from "../../RootComponent/RootComponent.js";
import styles from "./FlexItem.module.css";
const flexClassNames = {
    grow: styles.flexGrow,
    shrink: styles.flexShrink,
    content: styles.flexContent,
    fixed: styles.flexFixed
};
const alignSelfClassNames = {
    start: styles.alignSelfStart,
    end: styles.alignSelfEnd,
    center: styles.alignSelfCenter,
    baseline: styles.alignSelfBaseline,
    stretch: styles.alignSelfStretch
};
const warn = warnOnce('Flex.Item');
export const FlexItem = ({ children, alignSelf, flex, flexBasis, ...rest })=>{
    if (process.env.NODE_ENV === 'development') {
        warn('Компонент Flex.Item устарел, используйте компонент Flex в качестве альтернативы.');
    }
    return /*#__PURE__*/ _jsx(RootComponent, {
        ...rest,
        baseStyle: {
            flexBasis
        },
        baseClassName: classNames(alignSelf && alignSelfClassNames[alignSelf], flex && flexClassNames[flex]),
        children: children
    });
};
