// @vkontakte/vkui/dist/components/Search/Search.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { Icon16Clear, Icon16SearchOutline, Icon24Cancel } from "@vkontakte/icons";
import { classNames, hasReactNode, noop } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useAdaptivityConditionalRender } from "../../hooks/useAdaptivityConditionalRender/index.js";
import { useBooleanState } from "../../hooks/useBooleanState.js";
import { useConfigDirection } from "../../hooks/useConfigDirection.js";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { useNativeFormResetListener } from "../../hooks/useNativeFormResetListener.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { touchEnabled } from "../../lib/touch/index.js";
import { useIsomorphicLayoutEffect } from "../../lib/useIsomorphicLayoutEffect.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { Button } from "../Button/Button.js";
import { IconButton } from "../IconButton/IconButton.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { Headline } from "../Typography/Headline/Headline.js";
import { VisuallyHidden } from "../VisuallyHidden/VisuallyHidden.js";
const warn = warnOnce('Search');
/**
 * @see https://vkui.io/components/search
 */ export const Search = (_param)=>{
    var { className, getRootRef, style, placeholder: placeholderProp = 'Поиск', before = /*#__PURE__*/ _jsx(Icon16SearchOutline, {}), after = 'Отмена', getRef, icon: iconProp, onIconClick, autoComplete: autoCompleteProp = 'off', iconLabel, clearLabel = 'Очистить', clearButtonTestId, noPadding, findButtonText = 'Найти', findButtonTestId, onFindButtonClick, hideClearButton, slotProps } = _param, inputProps = _object_without_properties(_param, [
        "className",
        "getRootRef",
        "style",
        "placeholder",
        "before",
        "after",
        "getRef",
        "icon",
        "onIconClick",
        "autoComplete",
        "iconLabel",
        "clearLabel",
        "clearButtonTestId",
        "noPadding",
        "findButtonText",
        "findButtonTestId",
        "onFindButtonClick",
        "hideClearButton",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const direction = useConfigDirection();
    const isRtl = direction === 'rtl';
    const rootRest = useMergeProps({
        className,
        style,
        getRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const _useMergeProps = useMergeProps(_object_spread({
        getRootRef: getRef,
        placeholder: placeholderProp,
        autoComplete: autoCompleteProp,
        className: "vkuiSearch__nativeInput"
    }, inputProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input), { id, placeholder, onChange, autoComplete, getRootRef: getInputRef, onFocus: onInputFocus, onBlur: onInputBlur } = _useMergeProps, inputRest = _object_without_properties(_useMergeProps, [
        "id",
        "placeholder",
        "onChange",
        "autoComplete",
        "getRootRef",
        "onFocus",
        "onBlur"
    ]);
    const inputRef = useExternRef(getInputRef);
    const { value: isFocused, setTrue: setFocusedTrue, setFalse: setFocusedFalse } = useBooleanState(false);
    const generatedId = React.useId();
    const inputId = id ? id : `search-${generatedId}`;
    const [hasValue, setHasValue] = React.useState(()=>Boolean(inputRest.value || inputRest.defaultValue));
    const checkHasValue = (e)=>setHasValue(Boolean(e.currentTarget.value));
    const { sizeY = 'none' } = useAdaptivity();
    const { sizeY: adaptiveSizeY } = useAdaptivityConditionalRender();
    const platform = usePlatform();
    const hasAfter = platform === 'ios' && hasReactNode(after);
    const onFocus = (e)=>{
        setFocusedTrue();
        onInputFocus && onInputFocus(e);
    };
    const onBlur = (e)=>{
        setFocusedFalse();
        onInputBlur && onInputBlur(e);
    };
    const onCancel = React.useCallback(()=>{
        var _Object_getOwnPropertyDescriptor, _inputRef_current;
        // eslint-disable-next-line @typescript-eslint/unbound-method
        const nativeInputValueSetter = (_Object_getOwnPropertyDescriptor = Object.getOwnPropertyDescriptor(HTMLInputElement.prototype, 'value')) === null || _Object_getOwnPropertyDescriptor === void 0 ? void 0 : _Object_getOwnPropertyDescriptor.set;
        nativeInputValueSetter === null || nativeInputValueSetter === void 0 ? void 0 : nativeInputValueSetter.call(inputRef.current, '');
        const ev2 = new Event('input', {
            bubbles: true
        });
        (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.dispatchEvent(ev2);
    }, [
        inputRef
    ]);
    const onIconClickStart = React.useCallback((e)=>onIconClick === null || onIconClick === void 0 ? void 0 : onIconClick(e), [
        onIconClick
    ]);
    const onIconCancelClickStart = React.useCallback((e)=>{
        var _inputRef_current;
        e.preventDefault();
        (_inputRef_current = inputRef.current) === null || _inputRef_current === void 0 ? void 0 : _inputRef_current.focus();
        if (touchEnabled()) {
            onCancel();
        }
    }, [
        inputRef,
        onCancel
    ]);
    useIsomorphicLayoutEffect(()=>{
        if (inputRest.value !== undefined) {
            setHasValue(Boolean(inputRest.value));
        }
    }, [
        inputRest.value
    ]);
    useNativeFormResetListener(inputRef, ()=>{
        setHasValue(Boolean(inputRest.defaultValue));
    });
    const renderIconButton = (icon, props = {})=>/*#__PURE__*/ _jsxs(IconButton, _object_spread_props(_object_spread({
            hoverMode: "opacity",
            onPointerDown: onIconClickStart,
            className: "vkuiSearch__icon",
            onFocus: setFocusedTrue,
            onBlur: setFocusedFalse,
            onClick: noop
        }, props), {
            children: [
                /*#__PURE__*/ _jsx(VisuallyHidden, {
                    children: iconLabel
                }),
                icon
            ]
        }));
    const showControls = Boolean(iconProp || !hideClearButton || adaptiveSizeY.compact && onFindButtonClick);
    return /*#__PURE__*/ _jsxs(RootComponent, _object_spread_props(_object_spread({
        baseClassName: classNames('vkuiInternalSearch', "vkuiSearch__host", sizeY === 'none' && "vkuiSearch__sizeYNone", sizeY === 'compact' && "vkuiSearch__sizeYCompact", isFocused && "vkuiSearch__focused", hasValue && "vkuiSearch__hasValue", hasAfter && "vkuiSearch__hasAfter", iconProp && "vkuiSearch__hasIcon", inputRest.disabled && "vkuiSearch__disabled", !noPadding && "vkuiSearch__withPadding", isRtl && "vkuiSearch__rtl")
    }, rootRest), {
        children: [
            /*#__PURE__*/ _jsxs("div", {
                className: "vkuiSearch__field",
                children: [
                    /*#__PURE__*/ _jsx("label", {
                        htmlFor: inputId,
                        className: "vkuiSearch__label",
                        children: placeholder
                    }),
                    /*#__PURE__*/ _jsxs("div", {
                        className: "vkuiSearch__input",
                        children: [
                            before,
                            /*#__PURE__*/ _jsx(Headline, _object_spread({
                                Component: "input",
                                type: "search",
                                level: "1",
                                weight: "3",
                                id: inputId,
                                placeholder: placeholder,
                                autoComplete: autoComplete,
                                getRootRef: inputRef,
                                onChange: callMultiple(onChange, checkHasValue),
                                onFocus: onFocus,
                                onBlur: onBlur
                            }, inputRest))
                        ]
                    }),
                    showControls && /*#__PURE__*/ _jsxs("div", {
                        className: "vkuiSearch__controls",
                        children: [
                            iconProp && (typeof iconProp === 'function' ? iconProp(renderIconButton) : renderIconButton(iconProp)),
                            !hideClearButton && /*#__PURE__*/ _jsxs(IconButton, {
                                hoverMode: "opacity",
                                onPointerDown: onIconCancelClickStart,
                                onClick: onCancel,
                                className: "vkuiSearch__icon",
                                tabIndex: hasValue ? undefined : -1,
                                disabled: inputProps.disabled,
                                "data-testid": clearButtonTestId,
                                children: [
                                    /*#__PURE__*/ _jsx(VisuallyHidden, {
                                        children: clearLabel
                                    }),
                                    platform === 'ios' ? /*#__PURE__*/ _jsx(Icon16Clear, {}) : /*#__PURE__*/ _jsx(Icon24Cancel, {})
                                ]
                            }),
                            adaptiveSizeY.compact && onFindButtonClick && /*#__PURE__*/ _jsx(Button, {
                                mode: "primary",
                                size: "m",
                                className: classNames("vkuiSearch__findButton", adaptiveSizeY.compact.className),
                                focusVisibleMode: "inside",
                                onClick: onFindButtonClick,
                                tabIndex: hasValue ? undefined : -1,
                                "data-testid": findButtonTestId,
                                children: findButtonText
                            })
                        ]
                    })
                ]
            }),
            hasAfter && /*#__PURE__*/ _jsx("div", {
                className: "vkuiSearch__after",
                children: /*#__PURE__*/ _jsx(Button, {
                    mode: "tertiary",
                    size: "m",
                    focusVisibleMode: "inside",
                    hoverMode: "opacity",
                    activeMode: "opacity",
                    onClick: onCancel,
                    onFocus: setFocusedTrue,
                    onBlur: setFocusedFalse,
                    children: /*#__PURE__*/ _jsx("span", {
                        className: "vkuiSearch__afterTextClip",
                        children: after
                    })
                })
            })
        ]
    }));
};
