// @vkontakte/vkui/dist/components/ChipsInput/ChipsInput.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx } from "react/jsx-runtime";
import { useExternRef } from "../../hooks/useExternRef.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { ChipsInputBase } from "../ChipsInputBase/ChipsInputBase.js";
import { useChipsInput } from "./useChipsInput.js";
/**
 * @see https://vkui.io/components/chips-input
 */ export const ChipsInput = (_param)=>{
    var { // option
    value: valueProp, defaultValue, onChange, // input
    getRef, inputValue: inputValueProp, defaultInputValue: defaultInputValueProp, onInputChange: onInputChangeProp, getOptionValue, getOptionLabel, getNewOptionData, // other
    disabled: disabledProp, allowClearButton, delimiter, slotProps } = _param, restProps = _object_without_properties(_param, [
        "value",
        "defaultValue",
        "onChange",
        "getRef",
        "inputValue",
        "defaultInputValue",
        "onInputChange",
        "getOptionValue",
        "getOptionLabel",
        "getNewOptionData",
        "disabled",
        "allowClearButton",
        "delimiter",
        "slotProps"
    ]);
    const _useMergeProps = useMergeProps({
        getRootRef: getRef,
        value: inputValueProp,
        defaultValue: defaultInputValueProp,
        onChange: onInputChangeProp,
        disabled: disabledProp
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.input), { getRootRef: getInputRef, value: resolvedInputValue, defaultValue: resolvedInputDefaultValue, onChange: resolvedOnInputChange, disabled } = _useMergeProps, inputRest = _object_without_properties(_useMergeProps, [
        "getRootRef",
        "value",
        "defaultValue",
        "onChange",
        "disabled"
    ]);
    const { value, addOptionFromInput, removeOption, clearOptions, // input
    inputRef: inputRefHook, inputValue, onInputChange } = useChipsInput({
        // option
        value: valueProp,
        defaultValue,
        onChange,
        getOptionLabel,
        getOptionValue,
        getNewOptionData,
        // input
        inputValue: resolvedInputValue,
        defaultInputValue: resolvedInputDefaultValue,
        onInputChange: resolvedOnInputChange,
        // other
        disabled,
        delimiter
    });
    const inputRef = useExternRef(getInputRef, inputRefHook);
    return /*#__PURE__*/ _jsx(ChipsInputBase, _object_spread({
        value: value,
        clearButtonShown: allowClearButton && (!!value.length || !!inputValue.length),
        onAddChipOption: addOptionFromInput,
        onRemoveChipOption: removeOption,
        onClear: clearOptions,
        slotProps: _object_spread_props(_object_spread({}, slotProps), {
            input: _object_spread({
                getRootRef: inputRef,
                value: inputValue,
                onChange: onInputChange,
                disabled
            }, inputRest)
        })
    }, restProps));
};
