// @vkontakte/vkui/dist/components/Checkbox/CheckboxInput/CheckboxInput.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { Icon20CheckBoxIndetermanate, Icon20CheckBoxOff, Icon20CheckBoxOn, Icon24CheckBoxOff, Icon24CheckBoxOn } from "@vkontakte/icons";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivityConditionalRender } from "../../../hooks/useAdaptivityConditionalRender/index.js";
import { useExternRef } from "../../../hooks/useExternRef.js";
import { useMergeProps } from "../../../hooks/useMergeProps.js";
import { usePlatform } from "../../../hooks/usePlatform.js";
import { warnOnce } from "../../../lib/warnOnce.js";
import { RootComponent } from "../../RootComponent/RootComponent.js";
import { VisuallyHidden } from "../../VisuallyHidden/VisuallyHidden.js";
function setIndeterminate(el, indeterminate) {
    el.indeterminate = indeterminate;
}
const warn = warnOnce('Checkbox');
export function CheckboxInput(_param) {
    var { className, style, getRootRef, getRef, indeterminate, defaultIndeterminate, IconOnCompact = Icon20CheckBoxOn, IconOnRegular = Icon24CheckBoxOn, IconOffCompact = Icon20CheckBoxOff, IconOffRegular = Icon24CheckBoxOff, IconIndeterminate = Icon20CheckBoxIndetermanate, slotProps } = _param, restProps = _object_without_properties(_param, [
        "className",
        "style",
        "getRootRef",
        "getRef",
        "indeterminate",
        "defaultIndeterminate",
        "IconOnCompact",
        "IconOnRegular",
        "IconOffCompact",
        "IconOffRegular",
        "IconIndeterminate",
        "slotProps"
    ]);
    const rootRest = useMergeProps({
        className,
        style,
        getRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const _useMergeProps = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input), { onChange, getRootRef: getInputRef } = _useMergeProps, inputRest = _object_without_properties(_useMergeProps, [
        "onChange",
        "getRootRef"
    ]);
    const inputRef = useExternRef(getInputRef);
    const platform = usePlatform();
    const { sizeY: adaptiveSizeY } = useAdaptivityConditionalRender();
    React.useEffect(()=>{
        const indeterminateValue = indeterminate === undefined ? defaultIndeterminate : indeterminate;
        if (inputRef.current) {
            setIndeterminate(inputRef.current, Boolean(indeterminateValue));
        }
    }, [
        defaultIndeterminate,
        indeterminate,
        inputRef
    ]);
    const handleChange = React.useCallback((event)=>{
        if (defaultIndeterminate !== undefined && indeterminate === undefined && inputRest.checked === undefined && inputRef.current) {
            setIndeterminate(inputRef.current, false);
        }
        if (indeterminate !== undefined && inputRef.current) {
            setIndeterminate(inputRef.current, Boolean(indeterminate));
        }
        onChange && onChange(event);
    }, [
        defaultIndeterminate,
        indeterminate,
        inputRest.checked,
        onChange,
        inputRef
    ]);
    if (process.env.NODE_ENV === 'development') {
        /* istanbul ignore if: не проверяем в тестах */ if (getRef) {
            warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
        }
        if (defaultIndeterminate && inputRest.defaultChecked) {
            warn('defaultIndeterminate и defaultChecked не могут быть true одновременно', 'error');
        }
        if (indeterminate && inputRest.checked) {
            warn('indeterminate и checked не могут быть true одновременно', 'error');
        }
        if (inputRest.defaultChecked && inputRest.checked) {
            warn('defaultChecked и checked не могут быть true одновременно', 'error');
        }
    }
    return /*#__PURE__*/ _jsxs(RootComponent, _object_spread_props(_object_spread({
        baseClassName: "vkuiCheckboxInput__host"
    }, rootRest), {
        children: [
            /*#__PURE__*/ _jsx(VisuallyHidden, _object_spread({
                Component: "input",
                type: "checkbox",
                onChange: handleChange,
                getRootRef: inputRef,
                baseClassName: "vkuiCheckboxInput__input"
            }, inputRest)),
            platform === 'vkcom' ? /*#__PURE__*/ _jsx(IconOnCompact, {
                className: "vkuiCheckboxInput__iconOn"
            }) : /*#__PURE__*/ _jsxs(React.Fragment, {
                children: [
                    adaptiveSizeY.compact && /*#__PURE__*/ _jsx(IconOnCompact, {
                        className: classNames("vkuiCheckboxInput__iconOn", adaptiveSizeY.compact.className)
                    }),
                    adaptiveSizeY.regular && /*#__PURE__*/ _jsx(IconOnRegular, {
                        className: classNames("vkuiCheckboxInput__iconOn", adaptiveSizeY.regular.className)
                    })
                ]
            }),
            platform === 'vkcom' ? /*#__PURE__*/ _jsx(IconOffCompact, {
                className: "vkuiCheckboxInput__iconOff"
            }) : /*#__PURE__*/ _jsxs(React.Fragment, {
                children: [
                    adaptiveSizeY.compact && /*#__PURE__*/ _jsx(IconOffCompact, {
                        className: classNames("vkuiCheckboxInput__iconOff", adaptiveSizeY.compact.className)
                    }),
                    adaptiveSizeY.regular && /*#__PURE__*/ _jsx(IconOffRegular, {
                        className: classNames("vkuiCheckboxInput__iconOff", adaptiveSizeY.regular.className)
                    })
                ]
            }),
            platform === 'vkcom' ? /*#__PURE__*/ _jsx(IconIndeterminate, {
                width: 20,
                height: 20,
                className: "vkuiCheckboxInput__iconIndeterminate"
            }) : /*#__PURE__*/ _jsxs(React.Fragment, {
                children: [
                    adaptiveSizeY.compact && /*#__PURE__*/ _jsx(IconIndeterminate, {
                        className: classNames("vkuiCheckboxInput__iconIndeterminate", adaptiveSizeY.compact.className),
                        width: 20,
                        height: 20
                    }),
                    adaptiveSizeY.regular && /*#__PURE__*/ _jsx(IconIndeterminate, {
                        className: classNames("vkuiCheckboxInput__iconIndeterminate", adaptiveSizeY.regular.className),
                        width: 24,
                        height: 24
                    })
                ]
            })
        ]
    }));
}
