// @vkontakte/vkui/dist/components/Switch/Switch.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useConfigDirection } from "../../hooks/useConfigDirection.js";
import { useFocusVisible } from "../../hooks/useFocusVisible.js";
import { useFocusVisibleClassName } from "../../hooks/useFocusVisibleClassName.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { usePlatform } from "../../hooks/usePlatform.js";
import { callMultiple } from "../../lib/callMultiple.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { RootComponent } from "../RootComponent/RootComponent.js";
import { VisuallyHidden } from "../VisuallyHidden/VisuallyHidden.js";
const warn = warnOnce('Switch');
const sizeYClassNames = {
    none: "vkuiSwitch__sizeYNone",
    compact: "vkuiSwitch__sizeYCompact"
};
/**
 * @see https://vkui.io/components/switch
 */ export const Switch = (_param)=>{
    var { style: rootStyle, className: rootClassName, getRootRef: rootGetRootRef, getRef, slotProps } = _param, restProps = _object_without_properties(_param, [
        "style",
        "className",
        "getRootRef",
        "getRef",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const rootRest = useMergeProps({
        style: rootStyle,
        className: rootClassName,
        getRootRef: rootGetRootRef
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const _useMergeProps = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input), { checked: checkedProp, onBlur: onBlurProp, onFocus: onFocusProp, onClick } = _useMergeProps, inputRest = _object_without_properties(_useMergeProps, [
        "checked",
        "onBlur",
        "onFocus",
        "onClick"
    ]);
    const direction = useConfigDirection();
    const isRtl = direction === 'rtl';
    const platform = usePlatform();
    const { sizeY = 'none' } = useAdaptivity();
    const { focusVisible, onBlur, onFocus } = useFocusVisible();
    const focusVisibleClassNames = useFocusVisibleClassName({
        focusVisible,
        mode: 'outside'
    });
    const handleBlur = callMultiple(onBlur, onBlurProp);
    const handleFocus = callMultiple(onFocus, onFocusProp);
    const [localUncontrolledChecked, setLocalUncontrolledChecked] = React.useState(Boolean(inputRest.defaultChecked));
    const isControlled = checkedProp !== undefined;
    const syncUncontrolledCheckedStateOnClick = React.useCallback((e)=>{
        if (isControlled) {
            return;
        }
        const switchTarget = e.target;
        setLocalUncontrolledChecked(switchTarget.checked);
    }, [
        isControlled
    ]);
    const inputProps = _object_spread({
        Component: 'input',
        type: 'checkbox',
        role: 'switch',
        onBlur: handleBlur,
        onFocus: handleFocus,
        onClick: callMultiple(syncUncontrolledCheckedStateOnClick, onClick)
    }, inputRest);
    if (isControlled) {
        inputProps.checked = checkedProp;
        inputProps['aria-checked'] = checkedProp ? 'true' : 'false';
    } else {
        inputProps['aria-checked'] = localUncontrolledChecked ? 'true' : 'false';
    }
    return /*#__PURE__*/ _jsxs(RootComponent, _object_spread_props(_object_spread({
        Component: "label",
        baseClassName: classNames("vkuiSwitch__host", sizeY !== 'regular' && sizeYClassNames[sizeY], platform === 'ios' ? "vkuiSwitch__ios" : "vkuiSwitch__default", inputRest.disabled && "vkuiSwitch__disabled", isRtl && "vkuiSwitch__rtl", focusVisibleClassNames)
    }, rootRest), {
        children: [
            /*#__PURE__*/ _jsx(VisuallyHidden, _object_spread({
                baseClassName: "vkuiSwitch__inputNative"
            }, inputProps)),
            /*#__PURE__*/ _jsxs("span", {
                "aria-hidden": true,
                className: "vkuiSwitch__inputFake",
                children: [
                    /*#__PURE__*/ _jsx("span", {
                        className: "vkuiSwitch__track"
                    }),
                    /*#__PURE__*/ _jsx("span", {
                        "aria-hidden": true,
                        className: classNames("vkuiSwitch__handle", platform !== 'ios' && !inputRest.disabled && "vkuiSwitch__handleWithRipple")
                    })
                ]
            })
        ]
    }));
};
