// @vkontakte/vkui/dist/components/Radio/Radio.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { SelectionControl } from "../SelectionControl/SelectionControl.js";
import { SelectionControlLabel } from "../SelectionControl/SelectionControlLabel/SelectionControlLabel.js";
import { RadioInput } from "./RadioInput/RadioInput.js";
const warn = warnOnce('Radio');
/**
 * @see https://vkui.io/components/radio
 */ export const Radio = (_param)=>{
    var { children, description, style, className, getRootRef, titleAfter, getRef, labelProps, hoverMode, activeMode, hasHover, hasActive, focusVisibleMode, slotProps } = _param, restProps = _object_without_properties(_param, [
        "children",
        "description",
        "style",
        "className",
        "getRootRef",
        "titleAfter",
        "getRef",
        "labelProps",
        "hoverMode",
        "activeMode",
        "hasHover",
        "hasActive",
        "focusVisibleMode",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development') {
        if (labelProps) {
            warn('Свойство `labelProps` устаревшее, используйте `slotProps={ root: {...} }`');
        }
        if (getRef) {
            warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
        }
    }
    const rootProps = useMergeProps(_object_spread({
        style,
        className: classNames("vkuiRadio__host", className),
        getRootRef
    }, labelProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.root);
    const inputRest = useMergeProps(_object_spread({
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input);
    return /*#__PURE__*/ _jsxs(SelectionControl, _object_spread_props(_object_spread({
        hoverMode: hoverMode,
        activeMode: activeMode,
        hasHover: hasHover,
        hasActive: hasActive,
        focusVisibleMode: focusVisibleMode,
        disabled: inputRest.disabled
    }, rootProps), {
        children: [
            /*#__PURE__*/ _jsx(RadioInput, {
                slotProps: {
                    input: inputRest
                }
            }),
            /*#__PURE__*/ _jsx(SelectionControlLabel, {
                titleAfter: titleAfter,
                description: description,
                children: children
            })
        ]
    }));
};
Radio.Input = RadioInput;
