// @vkontakte/vkui/dist/components/Input/Input.js
'use client';
import { _ as _object_spread } from "@swc/helpers/_/_object_spread";
import { _ as _object_spread_props } from "@swc/helpers/_/_object_spread_props";
import { _ as _object_without_properties } from "@swc/helpers/_/_object_without_properties";
import { jsx as _jsx } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../hooks/useAdaptivity.js";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { FormField } from "../FormField/FormField.js";
import { UnstyledTextField } from "../UnstyledTextField/UnstyledTextField.js";
const warn = warnOnce('Input');
const sizeYClassNames = {
    none: "vkuiInput__sizeYNone",
    compact: "vkuiInput__sizeYCompact"
};
/**
 * @see https://vkui.io/components/input
 */ export const Input = (_param)=>{
    var { type = 'text', align = 'left', getRef, className: rootClassName, getRootRef, style, before, after, status, mode, slotProps } = _param, restProps = _object_without_properties(_param, [
        "type",
        "align",
        "getRef",
        "className",
        "getRootRef",
        "style",
        "before",
        "after",
        "status",
        "mode",
        "slotProps"
    ]);
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development' && getRef) {
        warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
    }
    const { sizeY = 'none' } = useAdaptivity();
    const _useMergeProps = useMergeProps({
        className: rootClassName,
        getRootRef,
        style
    }, slotProps === null || slotProps === void 0 ? void 0 : slotProps.root), { className } = _useMergeProps, rootProps = _object_without_properties(_useMergeProps, [
        "className"
    ]);
    const inputRest = useMergeProps(_object_spread({
        className: "vkuiInput__el",
        getRootRef: getRef
    }, restProps), slotProps === null || slotProps === void 0 ? void 0 : slotProps.input);
    return /*#__PURE__*/ _jsx(FormField, _object_spread_props(_object_spread({
        className: classNames("vkuiInput__host", align === 'right' && "vkuiInput__alignRight", align === 'center' && "vkuiInput__alignCenter", sizeY !== 'regular' && sizeYClassNames[sizeY], before && "vkuiInput__hasBefore", after && "vkuiInput__hasAfter", className),
        before: before,
        after: after,
        disabled: inputRest.disabled,
        mode: mode,
        status: status
    }, rootProps), {
        children: /*#__PURE__*/ _jsx(UnstyledTextField, _object_spread({
            as: "input",
            type: type
        }, inputRest))
    }));
};
