// @vkontakte/vkui/dist/cssm/components/CustomSelect/CustomSelectInput/CustomSelectInput.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
/* eslint-disable jsdoc/require-jsdoc */ import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useAdaptivity } from "../../../hooks/useAdaptivity.js";
import { useExternRef } from "../../../hooks/useExternRef.js";
import { useFocusWithin } from "../../../hooks/useFocusWithin.js";
import { useMergeProps } from "../../../hooks/useMergeProps.js";
import { usePlatform } from "../../../hooks/usePlatform.js";
import { getFormFieldModeFromSelectType } from "../../../lib/select.js";
import { FormField } from "../../FormField/FormField.js";
import { SelectTypography } from "../../SelectTypography/SelectTypography.js";
import { Text } from "../../Typography/Text/Text.js";
import { VisuallyHidden } from "../../VisuallyHidden/VisuallyHidden.js";
import styles from "./CustomSelectInput.module.css";
const sizeYClassNames = {
    none: styles.sizeYNone,
    compact: styles.sizeYCompact
};
/**
 * @since 5.10.0
 * @private
 */ export const CustomSelectInput = ({ style: rootStyle, className: rootClassName, getRootRef: rootGetRootRef, align = 'left', getRef, before, after, status, children, placeholder, selectType = 'default', multiline, fetching, labelTextTestId, searchable, accessible, slotProps, ...restProps })=>{
    const { sizeY = 'none' } = useAdaptivity();
    const { style, className, getRootRef, ...rootRest } = useMergeProps({
        style: rootStyle,
        className: rootClassName,
        getRootRef: rootGetRootRef
    }, slotProps?.root);
    const { className: inputClassName, value, readOnly, disabled, ...inputProps } = useMergeProps({
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    const title = children || placeholder;
    const showLabelOrPlaceholder = !Boolean(value);
    const handleRootRef = useExternRef(getRootRef);
    const focusWithin = useFocusWithin(handleRootRef);
    const inputReadonly = readOnly || disabled && fetching;
    const input = /*#__PURE__*/ _jsx(Text, {
        type: "text",
        disabled: disabled && !fetching,
        readOnly: inputReadonly,
        Component: "input",
        normalize: false,
        placeholder: children ? '' : placeholder,
        className: classNames(styles.el, (readOnly || showLabelOrPlaceholder && !focusWithin) && styles.elCursorPointer, inputClassName),
        value: value,
        ...inputProps
    });
    const inputHidden = React.useMemo(()=>{
        if (accessible) {
            if (!searchable) {
                return true;
            }
            return !focusWithin || inputReadonly && !fetching;
        } else {
            return false;
        }
    }, [
        accessible,
        fetching,
        focusWithin,
        inputReadonly,
        searchable
    ]);
    const labelHidden = showLabelOrPlaceholder ? false : !inputHidden;
    const platform = usePlatform();
    return /*#__PURE__*/ _jsx(FormField, {
        Component: "div",
        style: style,
        className: classNames(styles.host, align === 'right' && styles.alignRight, align === 'center' && styles.alignCenter, !children && styles.empty, multiline && styles.multiline, sizeY !== 'regular' && sizeYClassNames[sizeY], before && styles.hasBefore, after && styles.hasAfter, inputHidden && styles.inputHidden, labelHidden && styles.labelHidden, accessible && styles.accessible, value && styles.hasInputValue, className),
        getRootRef: handleRootRef,
        before: before,
        after: after,
        disabled: disabled,
        mode: getFormFieldModeFromSelectType(selectType),
        status: status,
        ...rootRest,
        children: /*#__PURE__*/ _jsxs("div", {
            className: styles.inputGroup,
            children: [
                /*#__PURE__*/ _jsx("div", {
                    className: classNames(styles.container, className),
                    tabIndex: -1,
                    "aria-hidden": true,
                    "data-testid": labelTextTestId,
                    children: /*#__PURE__*/ _jsx(SelectTypography, {
                        selectType: selectType,
                        className: styles.title,
                        children: title
                    })
                }),
                readOnly && platform === 'ios' ? /*#__PURE__*/ _jsx(VisuallyHidden, {
                    children: input
                }) : input
            ]
        })
    });
};
