// @vkontakte/vkui/dist/cssm/components/Radio/Radio.js
'use client';
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import * as React from "react";
import { classNames } from "@vkontakte/vkjs";
import { useMergeProps } from "../../hooks/useMergeProps.js";
import { warnOnce } from "../../lib/warnOnce.js";
import { SelectionControl } from "../SelectionControl/SelectionControl.js";
import { SelectionControlLabel } from "../SelectionControl/SelectionControlLabel/SelectionControlLabel.js";
import { RadioInput } from "./RadioInput/RadioInput.js";
import styles from "./Radio.module.css";
const warn = warnOnce('Radio');
/**
 * @see https://vkui.io/components/radio
 */ export const Radio = ({ children, description, style, className, getRootRef, titleAfter, getRef, labelProps, hoverMode, activeMode, hasHover, hasActive, focusVisibleMode, slotProps, ...restProps })=>{
    /* istanbul ignore if: не проверяем в тестах */ if (process.env.NODE_ENV === 'development') {
        if (labelProps) {
            warn('Свойство `labelProps` устаревшее, используйте `slotProps={ root: {...} }`');
        }
        if (getRef) {
            warn('Свойство `getRef` устаревшее, используйте `slotProps={ input: { getRootRef: ... } }`');
        }
    }
    const rootProps = useMergeProps({
        style,
        className: classNames(styles.host, className),
        getRootRef,
        ...labelProps
    }, slotProps?.root);
    const inputRest = useMergeProps({
        getRootRef: getRef,
        ...restProps
    }, slotProps?.input);
    return /*#__PURE__*/ _jsxs(SelectionControl, {
        hoverMode: hoverMode,
        activeMode: activeMode,
        hasHover: hasHover,
        hasActive: hasActive,
        focusVisibleMode: focusVisibleMode,
        disabled: inputRest.disabled,
        ...rootProps,
        children: [
            /*#__PURE__*/ _jsx(RadioInput, {
                slotProps: {
                    input: inputRest
                }
            }),
            /*#__PURE__*/ _jsx(SelectionControlLabel, {
                titleAfter: titleAfter,
                description: description,
                children: children
            })
        ]
    });
};
Radio.Input = RadioInput;
