// @vkontakte/vkjs/lib/device/index.js
export { IPHONE_SAFARI_BOTTOM_BAR } from "./IOSDetections.js";
export { IPHONE_X_SAFARI_BOTTOM_BAR } from "./IOSDetections.js";
export { IPHONE_KEYBOARD_REJECT_OFFSET } from "./IOSDetections.js";
export { IOS_NO_KEYBOARD_ALLOWED_OFFSET } from "./IOSDetections.js";
export { detectIOS } from "./IOSDetections.js";
export { isIPad } from "./IOSDetections.js";
export { isIPhone } from "./IOSDetections.js";
export { isIOS } from "./IOSDetections.js";
export { isIPadOS } from "./IOSDetections.js";
export { iosMajor } from "./IOSDetections.js";
export { iosMinor } from "./IOSDetections.js";
export { isWKWebView } from "./IOSDetections.js";
export { isScrollBasedViewport } from "./IOSDetections.js";
export { isIPhoneX } from "./IOSDetections.js";
export { isIOSChrome } from "./IOSDetections.js";
export { isLandscapePhone } from "./IOSDetections.js";
export { checkIPadOS } from "./IOSDetections.js";
export { hasMouse } from "./InputUtils.js";
export { hasHover } from "./InputUtils.js";
export { hasTouchEvents } from "./InputUtils.js";
export { hasTouch } from "./InputUtils.js";
export { isRetina } from "./retina.js";
